### ISSUE TITLE
Difference between `index_size = tab` and absence of `tab_width`

### ISSUE THREAD

**Issue by monnier** [editorconfig/editorconfig#508]
The spec says about `tab_width`:

> Set to a whole number defining the number of columns used to represent a tab character.
> This defaults to the value of `indent_size` and should not usually need to be specified.

But this suggests that

    indent_size = N

is equivalent to

    indent_size = tab
    tab_width = N

Is that really the intention?  This seems odd especially in the case where `indent_style = space` in which case TABs are presumably not used at all for the normal indentation and should thus have a display width that's not correlated to the indentation step.

**Comment by cxw42** [comment]
I don't think `tab_width` is meant to imply any other parameter.  Certainly it does not in [current editorconfig-vim](https://github.com/editorconfig/editorconfig-vim/blob/8b7da79e9daee7a3f3a8d4fe29886b9756305aff/plugin/editorconfig.vim#L449-L464).

**Comment by monnier** [comment]
My question is about the other direction: it says that just by setting `indent_size`, it also sets `tab_width`.

**Comment by xuhdev** [comment]
Yes, this is intended. If using space for indentation, the tab width is still used by many editors to control the effect of hitting the tab key.

**Comment by monnier** [comment]
But "the effect of hitting the tab key" is what is controlled by `indent_size`, not `tab_width`.  Those editors which conflate the two of course will set their "tab width" to the same value as `indent_size` but that doesn't explain why the EditorConfig standard needs to conflate the two.

**Comment by xuhdev** [comment]
@monnier 
Sorry for the confusion. `tab_width`, as it literally reads, means the size of a tab in the buffer that's being edited. The [spec](https://spec.editorconfig.org/) says:

> Set to a whole number defining the number of columns used to represent a tab character.

An editor uses it to control the appearance of a tab. It was incorrect for me to say it controls the effects of hitting `<Tab>` -- The effect is up to the editor itself, and it's related to `tab_width` only if hitting `<Tab>` outputs a tab character.

## The reason that we need `tab_width` to be default to `indent_size`

Consider the scenario that `indent_size` is set to `4`. 

- If the user uses tab as the indent character. By setting `tab_width` default to `indent_size`, the user doesn't need to adjust `tab_width` specifically.
- Otherwise, `tab_width` likely would not affect user's experience, as it is rare that `tab` is used in a source file if not for indentation purposes. A user can always manually set `tab_width` should a rare need arises.

**Comment by monnier** [comment]
> Consider the scenario that `indent_size` is set.
>
> - If the user uses tab as the indent character.

Do you mean by that `indent_style = tab`?

> By setting `tab_width` default to `indent_size`, the user doesn't need
> to adjust `tab_width` specifically.

But that's what

    indent_size = tab

is for, no?

> - Otherwise,

IOW when `indent_style = space`?

> `tab_width` likely would not affect user's experience,

That argument cuts both ways.  It can also be used in favor of the
choice that `tab_width` does not have to be set to `indent_size`.

**Comment by xuhdev** [comment]
On 6/20/24 08:09, monnier wrote:
>  > Consider the scenario that `indent_size` is set.
>  >
>  > - If the user uses tab as the indent character.
> 
> Do you mean by that `indent_style = tab`?

Yes.

> 
>  > By setting `tab_width` default to `indent_size`, the user doesn't need
>  > to adjust `tab_width` specifically.
> 
> But that's what
> 
> indent_size = tab
> 
> is for, no?

This is for indent_style = tab, and indent_size = a number.

> 
>  > - Otherwise,
> 
> IOW when `indent_style = space`?
> 
>  > `tab_width` likely would not affect user's experience,
> 
> That argument cuts both ways. It can also be used in favor of the
> choice that `tab_width` does not have to be set to `indent_size`.
> 

Yes, this bullet point is to clarify that it doesn't do harm when `indent_style` is space.

**Comment by cxw42** [comment]
@monnier Do you have an example of a bug or problem related to this issue?  E.g., is there a case where two plugins interpret one of these properties differently from each other?

**Comment by monnier** [comment]
> @monnier Do you have an example of a bug or problem related to this issue?

Yes: the problem is that it makes it almost impossible not to leave
`tab_width` at its standard/default value.  The problem there is that
this setting is not actually attached to those TABs, so as soon as code
is "disconnected" from the `.editorconfig` setting (either because you
look at it with a tool that does not pay attention to `.editorconfig`,
or because you copy the code "outside the reach" of `.editorconfig`,
such as when sending a code chunk by email) the code will tend to appear
(or behave, if the whitespace is significant) wrong because the tool
doesn't know the intended tag-width.= and will be forced to assume it
was set to the standard/default value.

**Comment by cxw42** [comment]
Hmm...

> this setting is not actually attached to those TABs

True, and this is a problem with any code that use a tabstop != 8 :( .  Whenever I open code that uses tab+space for indentation, but has tabstop != 8, it looks wrong.  I then try different tabstop values manually until I find one that looks right.  

Here's [a discussion I found via search](https://www.perlmonks.org/?node_id=847538) from July 2010 of just this issue.  You are certainly not alone in having this concern!

> "outside the reach" of `.editorconfig`,

... then there's nothing EditorConfig can do about it, by definition, unless I'm missing something :) .

**Comment by monnier** [comment]
>> "outside the reach" of `.editorconfig`,
>
> ... then there's nothing EditorConfig can do about it, by definition, unless
> I'm missing something :) .

It could try to refrain from encouraging non-standard values of
tab-width.

That's the whole purpose of this issue: only set `tab_width` if it is
*explicitly* set, rather than whenever one of `tab_width` or
`indent_size` is set (hence virtually always).

**Comment by cxw42** [comment]
> changing the default-behaviour of the current standard in a backward incompatible way is not applicable.

Yep, this is it.  There is currently no way for a `.editorconfig` file to say which spec version it uses.  Therefore, we cannot change the existing behaviour of `indent_size` without significant breakage.  Sorry!

**Comment by rearnsha** [comment]
Related to this, what is the intended value of tab_width for C files, given the following specification:
```
[*]
tab_width = 8
[*.c]
indent_size = 2
indent_style = tab
```
It seems that the implementation of editorconfig in Forgejo (which is written in Go) ignores the global tab_width setting and sets tab_width to 2, which is not what I might expect, but could be down to incorrect expectations on my part...

**Comment by monnier** [comment]
@rearnsha: Great example, thanks.  According to my reading of the spec, Forgejo got it wrong.
Not sure there's a right behavior, tho: maybe the best is to consider it as an inconsistent request and emit a warning/error.

**Comment by rearnsha** [comment]
@monnier thanks for confirming.  I've worked around the issue for now by explicitly setting tab_width whenever I have indent_style = tab and indent_size != 8, but I'd like to raise an issue with the devs and this will help.

https://forge.sourceware.org/gcc/gcc-TEST/commit/162ec539a39c0b24eea3e17863499aa3f89d6a64
